From 6420dae63a9c7836f15a8b1b2fccc458a5c8ba6f Mon Sep 17 00:00:00 2001
From: Michael Bestas <mkbestas@lineageos.org>
Date: Wed, 5 Sep 2018 19:22:34 +0300
Subject: [PATCH 5/9] sf: Workaround BufferLayer::setBuffers issue on 1440p
 msm8974

* msm8974 EGL blobs are reporting wrong supported dimensions,
  which is only an issue on devices with 1440p panels
  (find7 and lt03lte being the only known ones)
* Since we will not get any newer EGL blobs for msm8974,
  add an optional flag to skip the check that crashes
  surfaceflinger. Everything still operates normally like
  previous android versions.

Change-Id: I0ee5d97149713f5e024cd9f8014eca73d1ab2265
---
 services/surfaceflinger/Android.bp      | 7 +++++++
 services/surfaceflinger/BufferLayer.cpp | 2 ++
 2 files changed, 9 insertions(+)

diff --git a/services/surfaceflinger/Android.bp b/services/surfaceflinger/Android.bp
index 320e11f7e..c2a2292b9 100644
--- a/services/surfaceflinger/Android.bp
+++ b/services/surfaceflinger/Android.bp
@@ -155,6 +155,13 @@ cc_library_shared {
     lto: {
         thin: true,
     },
+    product_variables: {
+        lineage: {
+            msm8974_1440p_egl_workaround: {
+                cflags: ["-DALLOW_TOO_LARGE_DIMENSIONS"],
+            },
+        },
+    },
 }
 
 cc_binary {
diff --git a/services/surfaceflinger/BufferLayer.cpp b/services/surfaceflinger/BufferLayer.cpp
index 7ac143219..e6a6790a6 100644
--- a/services/surfaceflinger/BufferLayer.cpp
+++ b/services/surfaceflinger/BufferLayer.cpp
@@ -121,7 +121,9 @@ status_t BufferLayer::setBuffers(uint32_t w, uint32_t h, PixelFormat format, uin
     // can handle.
     if ((uint32_t(w) > maxSurfaceDims) || (uint32_t(h) > maxSurfaceDims)) {
         ALOGE("dimensions too large %u x %u", uint32_t(w), uint32_t(h));
+#ifndef ALLOW_TOO_LARGE_DIMENSIONS
         return BAD_VALUE;
+#endif
     }
 
     mFormat = format;
-- 
2.17.1

